### PR TITLE
Minor fixes for saving checkpoints.

### PR DESCRIPTION
**Pull request by csukuangfj**
Fixes https://github.com/k2-fsa/icefall/pull/259#issuecomment-1075884498

Let me test it first before merging.

### CODE REVIEW

**Comment by danpovey** [comment]
Fangjun, when running with this I get some odd effects.  I run with --epoch 14, but it prints "epoch 13" when training, plus a lot of warnings.
```
 python3 ./pruned_transducer_stateless2/train.py --exp-dir=pruned_transducer_stateless2/rework2f_sequential_lr10 --world-size 1 --num-epochs 20 --start-epoch 14 --\
full-libri 0 --max-duration 300 --lr-factor 10.0 --print-diagnostics=False&
[1] 3076060
de-74279-k2-train-6-0309102839-6644d5fdfd-gz27x:ASR: 
de-74279-k2-train-6-0309102839-6644d5fdfd-gz27x:ASR: 
de-74279-k2-train-6-0309102839-6644d5fdfd-gz27x:ASR: 2022-03-23 12:31:04,112 INFO [train.py:703] Training started
2022-03-23 12:31:04,163 INFO [train.py:713] Device: cuda:0
2022-03-23 12:31:04,168 INFO [train.py:722] {'best_train_loss': inf, 'best_valid_loss': inf, 'best_train_epoch': -1, 'best_valid_epoch': -1, 'batch_idx_train': 0, 'log_interval': 50, 'reset_interval': 200, 'valid_in\
terval': 1600, 'feature_dim': 80, 'subsampling_factor': 4, 'attention_dim': 512, 'nhead': 8, 'dim_feedforward': 2048, 'num_encoder_layers': 12, 'vgg_frontend': False, 'embedding_dim': 512, 'warm_step': 30000, 'model\
_warm_step': 3000, 'env_info': {'k2-version': '1.14', 'k2-build-type': 'Release', 'k2-with-cuda': True, 'k2-git-sha1': '6833270cb228aba7bf9681fccd41e2b52f7d984c', 'k2-git-date': 'Wed Mar 16 11:16:05 2022', 'lhotse-v\
ersion': '1.0.0.dev+git.e88c2a2.clean', 'torch-cuda-available': True, 'torch-cuda-version': '10.1', 'python-version': '3.8', 'icefall-git-branch': 'rework2f_sequential', 'icefall-git-sha1': '2bc317a-clean', 'icefall\
-git-date': 'Wed Mar 23 12:27:06 2022', 'icefall-path': '/ceph-dan/icefall', 'k2-path': '/ceph-dan/.local/lib/python3.8/site-packages/k2-1.14.dev20220317+cuda10.1.torch1.7.1-py3.8-linux-x86_64.egg/k2/__init__.py', '\
lhotse-path': '/ceph-dan/.local/lib/python3.8/site-packages/lhotse-1.0.0.dev0+git.e88c2a2.clean-py3.8.egg/lhotse/__init__.py', 'hostname': 'de-74279-k2-train-6-0309102839-6644d5fdfd-gz27x', 'IP address': '10.177.28.\
71'}, 'world_size': 1, 'master_port': 12354, 'tensorboard': True, 'num_epochs': 20, 'start_epoch': 14, 'start_batch': 0, 'exp_dir': PosixPath('pruned_transducer_stateless2/rework2f_sequential_lr10'), 'bpe_model': 'd\
ata/lang_bpe_500/bpe.model', 'lr_factor': 10.0, 'context_size': 2, 'prune_range': 5, 'lm_scale': 0.25, 'am_scale': 0.0, 'simple_loss_scale': 0.5, 'seed': 42, 'print_diagnostics': False, 'save_every_n': 8000, 'keep_l\
ast_k': 20, 'full_libri': False, 'manifest_dir': PosixPath('data/fbank'), 'max_duration': 300, 'bucketing_sampler': True, 'num_buckets': 30, 'concatenate_cuts': False, 'duration_factor': 1.0, 'gap': 1.0, 'on_the_fly\
_feats': False, 'shuffle': True, 'return_cuts': True, 'num_workers': 2, 'enable_spec_aug': True, 'spec_aug_time_warp_factor': 80, 'enable_musan': True, 'blank_id': 0, 'vocab_size': 500}
2022-03-23 12:31:04,168 INFO [train.py:724] About to create model
2022-03-23 12:31:14,123 INFO [train.py:728] Number of model parameters: 84440308
2022-03-23 12:31:14,147 INFO [checkpoint.py:100] Loading checkpoint from pruned_transducer_stateless2/rework2f_sequential_lr10/epoch-13.pt

de-74279-k2-train-6-0309102839-6644d5fdfd-gz27x:ASR: 2022-03-23 12:32:14,285 INFO [train.py:746] Loading optimizer state dict
2022-03-23 12:32:23,030 INFO [asr_datamodule.py:375] About to get train-clean-100 cuts
2022-03-23 12:34:07,516 INFO [asr_datamodule.py:199] Enable MUSAN
2022-03-23 12:34:07,516 INFO [asr_datamodule.py:200] About to get Musan cuts
2022-03-23 12:34:44,103 INFO [asr_datamodule.py:228] Enable SpecAugment
2022-03-23 12:34:44,103 INFO [asr_datamodule.py:229] Time warp factor: 80
2022-03-23 12:34:44,104 INFO [asr_datamodule.py:241] Num frame mask: 10
2022-03-23 12:34:44,104 INFO [asr_datamodule.py:254] About to create train dataset
2022-03-23 12:34:44,104 INFO [asr_datamodule.py:282] Using BucketingSampler.
2022-03-23 12:34:46,985 INFO [asr_datamodule.py:298] About to create train dataloader
2022-03-23 12:34:46,985 INFO [asr_datamodule.py:301] Loading sampler state dict
/ceph-dan/.local/lib/python3.8/site-packages/lhotse-1.0.0.dev0+git.e88c2a2.clean-py3.8.egg/lhotse/dataset/sampling/simple.py:144: UserWarning: SimpleCutSampler.load_state_dict(): Inconsistent time_constraint:
expected TimeConstraint(max_duration=300, max_samples=None, max_frames=None, current=0, num_cuts=0)
received TimeConstraint(max_duration=300, max_samples=None, max_frames=None, current=142.86206250000004, num_cuts=34)
We will overwrite the settings with the received state_dict.
/ceph-dan/.local/lib/python3.8/site-packages/lhotse-1.0.0.dev0+git.e88c2a2.clean-py3.8.egg/lhotse/dataset/sampling/simple.py:144: UserWarning: SimpleCutSampler.load_state_dict(): Inconsistent time_constraint:
expected TimeConstraint(max_duration=300, max_samples=None, max_frames=None, current=0, num_cuts=0)
received TimeConstraint(max_duration=300, max_samples=None, max_frames=None, current=58.650812499999994, num_cuts=8)
<more warnings, snip>
expected TimeConstraint(max_duration=300, max_samples=None, max_frames=None, current=0, num_cuts=0)
received TimeConstraint(max_duration=300, max_samples=None, max_frames=None, current=87.55556250000001, num_cuts=5)
We will overwrite the settings with the received state_dict.
/ceph-dan/.local/lib/python3.8/site-packages/lhotse-1.0.0.dev0+git.e88c2a2.clean-py3.8.egg/lhotse/dataset/sampling/simple.py:144: UserWarning: SimpleCutSampler.load_state_dict(): Inconsistent time_constraint:
expected TimeConstraint(max_duration=300, max_samples=None, max_frames=None, current=0, num_cuts=0)
received TimeConstraint(max_duration=300, max_samples=None, max_frames=None, current=182.68893749999998, num_cuts=10)
We will overwrite the settings with the received state_dict.
2022-03-23 12:34:49,982 INFO [asr_datamodule.py:396] About to get dev-clean cuts
2022-03-23 12:34:53,166 INFO [asr_datamodule.py:401] About to get dev-other cuts
2022-03-23 12:34:55,415 INFO [asr_datamodule.py:323] About to create dev dataset
/ceph-dan/.local/lib/python3.8/site-packages/lhotse-1.0.0.dev0+git.e88c2a2.clean-py3.8.egg/lhotse/dataset/sampling/bucketing.py:96: UserWarning: Lazy CutSet detected in BucketingSampler: this is not well supported y\
et, and you might experience a potentially long lag while the buckets are being created.
2022-03-23 12:34:55,564 INFO [asr_datamodule.py:342] About to create dev dataloader
2022-03-23 12:34:55,565 INFO [train.py:848] Sanity check -- see if any of the batches in epoch 0 would cause OOM.
/ceph-dan/.local/lib/python3.8/site-packages/lhotse-1.0.0.dev0+git.e88c2a2.clean-py3.8.egg/lhotse/dataset/sampling/utils.py:71: UserWarning: Empty sampler encountered in find_pessimistic_batches()
2022-03-23 12:34:55,566 INFO [train.py:804] epoch 13, learning rate 0.0019382104649663236
/ceph-dan/.local/lib/python3.8/site-packages/torchaudio/backend/utils.py:53: UserWarning: "sox" backend is being deprecated. The default backend will be changed to "sox_io" backend in 0.8.0 and "sox" backend will be\
 removed in 0.9.0. Please migrate to "sox_io" backend. Please refer to https://github.com/pytorch/audio/issues/903 for the detail.
  warnings.warn(
/ceph-dan/.local/lib/python3.8/site-packages/torchaudio/backend/utils.py:53: UserWarning: "sox" backend is being deprecated. The default backend will be changed to "sox_io" backend in 0.8.0 and "sox" backend will be\
 removed in 0.9.0. Please migrate to "sox_io" backend. Please refer to https://github.com/pytorch/audio/issues/903 for the detail.
  warnings.warn(
2022-03-23 12:35:11,882 INFO [train.py:643] Epoch 13, batch 0, loss[loss=0.171, simple_loss=0.2781, pruned_loss=0.03193, over 7317.00 frames.], tot_loss[loss=0.171, simple_loss=0.2781, pruned_loss=0.03193, over 7317\
.00 frames.], batch size: 22
2022-03-23 12:38:51,037 INFO [train.py:643] Epoch 13, batch 50, loss[loss=0.18, simple_loss=0.2799, pruned_loss=0.04003, over 7463.00 frames.], tot_loss[loss=0.176, simple_loss=0.2715, pruned_loss=0.04025, over 3317\
34.87 frames.], batch size: 26
2022-03-23 12:42:20,733 INFO [train.py:643] Epoch 13, batch 100, loss[loss=0.1819, simple_loss=0.2858, pruned_loss=0.03902, over 7213.00 frames.], tot_loss[loss=0.1759, simple_loss=0.2722, pruned_loss=0.03977, over \
582114.16 frames.], batch size: 24
2022-03-23 12:45:05,650 INFO [tr
```

**Comment by danpovey** [comment]
... Another oddity is that the pruned_loss is quite a bit different when it's rerun from the checkpoint vs. the first time I run it.  Originally it was around 0.05, now it's about 0.04.

**Comment by csukuangfj** [comment]
You are using `--start-epoch 14` and it indeed loaded the checkpoint `epoch-13.pt`. 

But it's strange that the log shows `Epoch 13, batch 0`, I cannot reproduce it locally.
The following is what I get when using `--start-epoch 14`:

```
2022-03-23 13:50:35,299 INFO [checkpoint.py:100] Loading checkpoint from pruned_transducer_stateless/exp-2/epoch-13.pt       [30/1945]
2022-03-23 13:50:39,999 INFO [asr_datamodule.py:375] About to get train-clean-100 cuts
2022-03-23 13:50:46,271 INFO [train.py:797] Before removing short and long utterances: 85617
2022-03-23 13:50:46,271 INFO [train.py:798] After removing short and long utterances: 85611
2022-03-23 13:50:46,271 INFO [train.py:799] Removed 6 utterances (0.00701%)
2022-03-23 13:50:46,271 INFO [asr_datamodule.py:199] Enable MUSAN
2022-03-23 13:50:46,271 INFO [asr_datamodule.py:200] About to get Musan cuts
2022-03-23 13:50:48,946 INFO [asr_datamodule.py:228] Enable SpecAugment
2022-03-23 13:50:48,946 INFO [asr_datamodule.py:229] Time warp factor: 80
2022-03-23 13:50:48,946 INFO [asr_datamodule.py:241] Num frame mask: 10
2022-03-23 13:50:48,946 INFO [asr_datamodule.py:254] About to create train dataset
2022-03-23 13:50:48,946 INFO [asr_datamodule.py:282] Using BucketingSampler.
2022-03-23 13:50:49,362 INFO [asr_datamodule.py:298] About to create train dataloader
2022-03-23 13:50:49,362 INFO [asr_datamodule.py:396] About to get dev-clean cuts
2022-03-23 13:50:49,506 INFO [asr_datamodule.py:401] About to get dev-other cuts
2022-03-23 13:50:49,620 INFO [asr_datamodule.py:323] About to create dev dataset
2022-03-23 13:50:49,629 INFO [asr_datamodule.py:342] About to create dev dataloader
2022-03-23 13:50:49,630 INFO [train.py:874] Sanity check -- see if any of the batches in epoch 0 would cause OOM.
/ceph-fj/fangjun/open-source-2/lhotse-rnnt-ali/lhotse/dataset/sampling/simple.py:237: UserWarning: The first cut drawn in batch collec
tion violates the max_frames, max_cuts, or max_duration constraints - we'll return it anyway. Consider increasing max_frames/max_cuts/
max_duration.
  warnings.warn(
/ceph-fj/fangjun/open-source-2/icefall-master/egs/librispeech/ASR/pruned_transducer_stateless/conformer.py:113: UserWarning: __floordi
v__ is deprecated, and its behavior will change in a future version of pytorch. It currently rounds toward 0 (like the 'trunc' functio
n NOT 'floor'). This results in incorrect rounding for negative values. To keep the current behavior, use torch.div(a, b, rounding_mod
e='trunc'), or for actual floor division, use torch.div(a, b, rounding_mode='floor').
  lengths = ((x_lens - 1) // 2 - 1) // 2
./pruned_transducer_stateless/train.py:496: UserWarning: __floordiv__ is deprecated, and its behavior will change in a future version
of pytorch. It currently rounds toward 0 (like the 'trunc' function NOT 'floor'). This results in incorrect rounding for negative valu
es. To keep the current behavior, use torch.div(a, b, rounding_mode='trunc'), or for actual floor division, use torch.div(a, b, roundi
ng_mode='floor').
  lengths = ((x_lens - 1) // 2 - 1) // 2
./pruned_transducer_stateless/train.py:496: UserWarning: __floordiv__ is deprecated, and its behavior will change in a future version
of pytorch. It currently rounds toward 0 (like the 'trunc' function NOT 'floor'). This results in incorrect rounding for negative valu
es. To keep the current behavior, use torch.div(a, b, rounding_mode='trunc'), or for actual floor division, use torch.div(a, b, roundi
ng_mode='floor').
  info["frames"] = (feature_lens // params.subsampling_factor).sum().item()
2022-03-23 13:50:54,480 INFO [train.py:834] epoch 14, learning rate 2.551551815399144e-07
2022-03-23 13:50:55,023 INFO [train.py:672] Epoch 14, batch 0, loss[loss=0.1929, simple_loss=0.2558, pruned_loss=0.06503, over 378.00
frames.], tot_loss[loss=0.1929, simple_loss=0.2558, pruned_loss=0.06503, over 378.00 frames.], batch size: 1
2022-03-23 13:51:00,257 INFO [train.py:672] Epoch 14, batch 50, loss[loss=0.1187, simple_loss=0.2271, pruned_loss=0.005102, over 400.00 frames.], tot_loss[loss=0.1598, simple_loss=0.2538, pruned_loss=0.03289, over 14888.84 frames.], batch size: 1

**Comment by csukuangfj** [comment]
> Originally it was around 0.05, now it's about 0.04.

Originally, we did not save the sampler's state dict when saving checkpoint.

You can set `sampler_state_dict = None` in the following code to recover the previous behavior, though I am not sure whether it is caused by that.
https://github.com/k2-fsa/icefall/blob/6a091da0b0543befb0492848d3583700c274d111/egs/librispeech/ASR/pruned_transducer_stateless/train.py#L801-L808

**Comment by danpovey** [comment]
It is doing
`    params["start_epoch"] = saved_params["cur_epoch"]`
which would affect this...

**Comment by csukuangfj** [comment]
Thanks, I will propose a fix.

**Comment by danpovey** [comment]
OK, the reason for the loss being better than expected seems to be that we are re-using the same random seed we used for the epoch we just trained the model on, so the data is more similar to previous data than it would otherwise be.  (In particular, the spec-aug masks will be in the same places).

**Comment by pzelasko** [comment]
I can add RNG state to spec augment state dict to avoid this, but you’ll need to save/load it manually. Might be worth to save global random states from torch, numpy and Python and restore them too.

**Comment by danpovey** [comment]
Well this happened because of re-using the exact same seed that we had already trained on, due to a bug in setting the epoch that has since been fixed. I don't think it would be harmful if we happened to use a different seed because of restarting from a checkpoint.
